common/clock/ratelimiter_test.go

```diff
@@ -0,0 +1,733 @@
+						case 3:
+							// call Reserve on everything, and cancel half of them
+							now := time.Now() // needed for the real ratelimiter to cancel successfully like the wrapper does
+							_a, _w, _m := actual.ReserveN(now, 1), wrapped.Reserve(), mocked.Reserve()
+							if rng.Intn(2)%2 == 0 {
+								_a.CancelAt(now)
+								_w.Used(false)
+								_m.Used(false)
+								t.Logf("round[%v][%v] ReserveWithCancel, canceled", round, event)
+								compressedReplay[round][event] = func(t *testing.T) {
+									compressed.Reserve().Used(false)
+									t.Logf("compressed round[%v][%v] ReserveWithCancel, canceled", round, event)
+								}
+							} else {
+								a, w, m := _a.OK() && _a.Delay() == 0, _w.Allow(), _m.Allow()
+								check(t, "ReserveWithCancel", round, event, a, w, m, nil)
+								if !a {
+									// must cancel, or the not-yet-available reservation affects future calls.
+									// that's valid if you intend to wait for your reservation, but the wrapper
+									// does not allow that.
+									_a.Cancel()
+								}
+								_w.Used(w)
+								_m.Used(m)
+								compressedReplay[round][event] = func(t *testing.T) {
+									_c := compressed.Reserve()
+									c := _c.Allow()
+									_c.Used(c)
+									check(t, "ReserveWithCancel (Compressed)", round, event, a, w, m, &c)
+								}
+							}
+						}
+					}
+					debug(t, "Tokens after round, real: %0.2f, wrapped: %0.2f, mocked: %0.2f", actual.Tokens(), wrapped.Tokens(), mocked.Tokens())
+				}
+			}
+			t.Run("compressed time", func(t *testing.T) {
+				// and now replay the compressed ratelimiter and make sure it matches too,
+				// as time-compressed must behave the same as real-time.
+				//
+				// this is primarily intended to detect cases where real-time is accidentally used,
+				// as ~zero time actually passes, which is quite different from the above tests.
+				//
+				// it's not perfect, but it does eventually notice such bugs, as you can see by
+				// changing literally any timesource.Now() calls into time.Now() and running
+				// tests a few times.
+				// depending on the change it might not notice in most tests, but eventually a
+				// problematic combination is triggered and can be replayed with the logged seed.
+				for round := range calls {
+					for event := range calls[round] {
+						compressedTS.Advance(granularity)
+						debug(t, "Tokens before compressed round: %0.2f", compressed.Tokens())
+						replay := compressedReplay[round][event]
+						if replay != nil {
+							replay(t)
+						}
+						debug(t, "Tokens after compressed round: %0.2f", compressed.Tokens())
+					}
+				}
+			})
+		})
+	}
+}
+
+func maxDur(d time.Duration, ds ...time.Duration) time.Duration {
+	for _, tmp := range ds {
+		if tmp > d {
+			d = tmp
+		}
+	}
+	return d
+}
+func minDur(d time.Duration, ds ...time.Duration) time.Duration {
+	for _, tmp := range ds {
+		if tmp < d {
+			d = tmp
+		}
+	}
+	return d
+}
+
+type timesourceContext struct {
```

Comment:
possibly useful to break out separately: context-timeouts need a timesource-wrapper.